### PR TITLE
Better websockets

### PR DESCRIPTION
**Pull request by provinzkraut**
Add enhanced websocket support, making websocket handling a lot easier and feel more "Starlite".

The main features are two new classes:

1. The `websocket_listener`: A special websocket handler, listening and sending incoming data in a loop
2. `WebsocketListener`: An abstract base class building upon `websocket_listener` and allowing to define things in a more OOP way

<hr>

## Examples

```python
from starlite.handlers.websocket_handlers import websocket_listener

@websocket_listener("/")
async def handler(data: str) -> str:
  return data
```

this is equivalent to:

```python
from starlite import websocket, WebSocket
from starlite.exceptions import WebSocketDisconnect


@websocket("/")
async def handler(socket: WebSocket) -> str:
  await socket.accept()
  while True:
    try:
      data = await socket.receive_text()
      await socket.send_text(data)
    except WebSocketDisconnect:
      break
```

It allows to send and receive either bytes, text or json, all configure by the type annotations of the handler function:

```python
from starlite.handlers.websocket_handlers import websocket_listener

@websocket_listener("/")
async def handler(data: list[str]) -> dict[str, str]:
  return {"message": data}
```

### Dependencies

Dependencies are handled on the route handler level, and passed into the listener function, so this works as expected:

```python
from starlite.di import Provide
from starlite.handlers.websocket_handlers import websocket_listener


@websocket_listener("/", dependencies={"foo": Provide(lambda: "foo")})
async def handler(data: str, foo: str) -> str:
  return data
```

And since this uses the regular signature/kwargs modelling, we can still inject things like `socket` or `state`:

```python
from starlite.handlers.websocket_handlers import websocket_listener
from starlite.datastructures import State

@websocket_listener("/")
async def handler(data: str, socket: WebSocket, state: State) -> str:
  return data
```

### Handling websockets synchronously

Since we don't need to handle the socket directly anymore, it's possible to use synchronous functions as well and this "just works":

```python
from starlite.handlers.websocket_handlers import websocket_listener

@websocket_listener("/")
def handler(data: str) -> str:
  return data
```

### Class based listeners

The class based listener build upon the `websocket_listener`, by providing an OOP way to handle websocket events, using synchronous or asynchronous methods as well:

```python
from starlite.handlers.websocket_handlers import WebsocketListener

class MyListener(WebsocketListener):
  path = "/"

  def on_connect(self, socket: WebSocket) -> None:
    print("socket connected")

  def on_receive(self, data: str) -> str:
    return "hello, world!"
  
  def on_disconnect(self, socket: WebSocket) -> None:
    print("socket disconnected")
```

## Todo

Ideally we want to integrate similar functionalities to route handlers in regards of data (de)serialisation. This is probably best done once #1295 is merged, as it would greatly simplify this. 

This is not a blocker though, and from a functionality standpoint this can be released as-is.

### CODE REVIEW

**Comment by v3ss0n** [comment]
i would be start testing this in a few days , really nice.

**Comment by julio-34727** [comment]
Hi @provinzkraut , will there be a way with the new websocket to emit to frontend out of the route handle like in flask_socketio (https://flask-socketio.readthedocs.io/en/latest/) ?

**Comment by provinzkraut** [comment]
> Hi @provinzkraut , will there be a way with the new websocket to emit to frontend out of the route handle like in flask_socketio (https://flask-socketio.readthedocs.io/en/latest/) ?

Can you be a bit more specific which of their features you're referring to?

Broadcasting (sending a message to all connected clients) won't be a built-in functionality implemented here, but it's planned. You can look at a very basic implementation of this here: https://github.com/starlite-api/starlite/pull/1378

**Review by Goldziher — COMMENTED** [review]
Very nice all in all

File: starlite/handlers/websocket_handlers.py
Comment (provinzkraut) [reply starlite/handlers/websocket_handlers.py:182]: Because no callable with the correct signature exists. 

There is the callback, which has the additional dependencies and `data` but not necessarily `socket`, and the handler function we create, which has `socket` but none of the additional dependencies defined by the callback.

So we need to create a signature from a function that looks exactly like the callback, minus `data`, and which is guaranteed to include `socket`.

Maybe I don't follow what you mean, but I don't think your suggestion would accomplish that?

Comment (Goldziher) [reply starlite/handlers/websocket_handlers.py:182]: Ok, lets revisit this code again in the future for refactoring - i think we can address this in a more cohesive way with some of the approaches that have been recently introduced by @peterschutt, but its not a blocker.

File: starlite/handlers/websocket_handlers.py
Comment (provinzkraut) [inline starlite/handlers/websocket_handlers.py:211]: I'm not sure if we should do this runtime validation, as we don't have this in place everywhere else. @starlite-api/maintainers?

File: starlite/handlers/websocket_handlers.py
Comment (provinzkraut) [inline starlite/handlers/websocket_handlers.py:216]: See comment above

File: starlite/handlers/websocket_handlers.py
Comment (Goldziher) [reply starlite/handlers/websocket_handlers.py:211]: im not in favor

File: docs/examples/websockets/dependency_injection_yield.py
Comment (Goldziher) [inline docs/examples/websockets/dependency_injection_yield.py:1]: can you rebase this PR on the litestar stuff?

File: starlite/handlers/websocket_handlers.py
Comment (Goldziher) [inline starlite/handlers/websocket_handlers.py:184]: ?

Comment (Goldziher) [reply starlite/handlers/websocket_handlers.py:184]: why is this method empty?

File: starlite/handlers/websocket_handlers.py
Comment (Goldziher) [inline starlite/handlers/websocket_handlers.py:347]: why not mark as abstract?

File: starlite/handlers/websocket_handlers.py
Comment (provinzkraut) [reply starlite/handlers/websocket_handlers.py:211]: Then let's not

File: starlite/handlers/websocket_handlers.py
Comment (provinzkraut) [reply starlite/handlers/websocket_handlers.py:184]: Because the validation needs to happen elsewhere and the validation of the base classes for this needs to be disabled since it does not apply in this case.

File: starlite/handlers/websocket_handlers.py
Comment (provinzkraut) [reply starlite/handlers/websocket_handlers.py:362]: Because then I'd have to check if this method has been implemented. By not making it abstract I can just use it as-is.

File: docs/examples/websockets/dependency_injection_yield.py
Comment (provinzkraut) [reply docs/examples/websockets/dependency_injection_yield.py:1]: I did.

**Comment by provinzkraut** [comment]
I rebased this on, and set the target to `rename-target`. Once that one is merged, this PR can be merged into `main`.

File: starlite/handlers/websocket_handlers.py
Comment (Goldziher) [reply starlite/handlers/websocket_handlers.py]: so can you add a comment explaining this?

**Review by Goldziher — APPROVED** [review]

File: starlite/handlers/websocket_handlers.py
Comment (provinzkraut) [reply starlite/handlers/websocket_handlers.py]: Okay so the base class of this is `WebSocketRouteHandler`, which performs a bunch of validations, all of which do not apply to this subclass. For example, it checks that handler functions return `None`, have to be asynchronous, need to have a `socket` parameter, etc. All of this is not applicable in this case, so we just disable this completely.

We cannot however use this method to validate the handler passed in by the user, because that's already been wrapped by the time `_validate_handler_function` is called, since this happens *after* it has been replaced by the handler instance; Upon being registered on the application instance.

Comment (provinzkraut) [reply starlite/handlers/websocket_handlers.py]: Does that make sense?

Comment (Goldziher) [reply starlite/handlers/websocket_handlers.py]: i got it, i meant - add a comment in the code 😉

Comment (provinzkraut) [reply starlite/handlers/websocket_handlers.py]: Oh :cry:

**Pull request merged**
